src/uct/cuda/cuda_ipc/cuda_ipc.cuh

```diff
@@ -0,0 +1,216 @@
+/**
+* Copyright (c) NVIDIA CORPORATION & AFFILIATES, 2025. ALL RIGHTS RESERVED.
+*
+* See file LICENSE for terms.
+*/
+
+#ifndef UCT_CUDA_IPC_CUH
+#define UCT_CUDA_IPC_CUH
+
+#include "ucs/type/status.h"
+#include "uct/api/uct_def.h"
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include <uct/api/device/uct_device_types.h>
+
+extern "C" {
+#include <uct/cuda/cuda_ipc/cuda_ipc_device.h>
+}
+
+#define align_pow2(_n, _p) (!((_n) & ((_p) - 1)))
+#define WARP_SIZE 32
+#define COPY_LOOP_UNROLL 8
+
+__device__ __forceinline__ int4 ld_global_cg(const int4* p) {
+    int4 v;
+    asm volatile ("ld.global.cg.v4.s32 {%0,%1,%2,%3}, [%4];"
+                  : "=r"(v.x), "=r"(v.y), "=r"(v.z), "=r"(v.w)
+                  : "l"(p));
+    return v;
+}
+
+__device__ __forceinline__ void st_global_cg(int4* p, const int4& v) {
+    asm volatile ("st.global.cg.v4.s32 [%0], {%1,%2,%3,%4};"
+                  :
+                  : "l"(p), "r"(v.x), "r"(v.y), "r"(v.z), "r"(v.w));
+}
+
+__device__ __forceinline__ int2 ld_global_cg(const int2* p) {
+    int2 v;
+    asm volatile ("ld.global.cg.v2.s32 {%0,%1}, [%2];"
+                  : "=r"(v.x), "=r"(v.y)
+                  : "l"(p));
+    return v;
+}
+
+__device__ __forceinline__ void st_global_cg(int2* p, const int2& v) {
+    asm volatile ("st.global.cg.v2.s32 [%0], {%1,%2};"
+                  :
+                  : "l"(p), "r"(v.x), "r"(v.y));
+}
+
+__device__ inline void uct_cuda_ipc_copy_single_nv(void *dst, const void *src,
+                                                   size_t size)
+{
+    auto s1 = reinterpret_cast<const char*>(src);
+    auto d1 = reinterpret_cast<char *>(dst);
+
+    for (size_t i = threadIdx.x; i <size; i += blockDim.x) {
+        d1[i] = s1[i];
+    }
+}
+
+template<int UNROLL>
+__device__ inline void uct_cuda_ipc_copy_single(void *dst, const void *src,
+                                                size_t size)
+{
+    using vec4 = int4;
+    using vec2 = int2;
+    auto s1  = reinterpret_cast<const char*>(src);
+    auto d1  = reinterpret_cast<char *>(dst);
+    const vec4 *s4;
+    vec4 *d4;
+    int warp, num_warps, idx;
+    size_t num_lines;
+
+    if (align_pow2((intptr_t)s1, sizeof(vec4)) && align_pow2((intptr_t)d1, sizeof(vec4))) {
+        vec4 tmp[UNROLL];
+        warp      = threadIdx.x / WARP_SIZE;
+        num_warps = blockDim.x / WARP_SIZE;
+        idx       = threadIdx.x % WARP_SIZE;
```

Comment:
We will probably need similar logic for other transports. We did something similar in the POC (look for `uct_gdaki_exec_init` in the POC fork repo).
I will place it in a common file with other useful macros and assembly funcs